### PR TITLE
added contact types

### PR DESCRIPTION
**Pull request by stemod**

### CODE REVIEW

File: Controller/ContactController.php [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+use Sulu\Component\Rest\ListBuilder\Doctrine\FieldDescriptor\DoctrineFieldDescriptor;
+use Sulu\Component\Rest\ListBuilder\ListRepresentation;
+use Sulu\Component\Rest\RestHelperInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+class ContactController extends SuluContactController
+{
+    /**
+     * lists all contacts
+     * optional parameter 'flat' calls listAction.
```
Comment (turbo-ele) [inline Controller/ContactController.php]: List all contacts.
Optional parameter 'flat' calls listAction.

File: Controller/ContactController.php [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+     *
+     * @param Request $request
+     *
+     * @return \Symfony\Component\HttpFoundation\Response
```
Comment (turbo-ele) [inline Controller/ContactController.php]: i'd prefer

```
@return Response
```

 and adding an import

File: Controller/ContactController.php [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            } else {
+                $contacts = $this->getDoctrine()->getRepository(
+                    $this->container->getParameter('sulu.model.contact.class')
+                )->findAll();
+                $serializationGroups = array_merge(
+                    $serializationGroups,
+                    static::$contactSerializationGroups
+                );
+            }
+            // convert to api-contacts
```
Comment (turbo-ele) [inline Controller/ContactController.php]: format

File: Controller/ContactController.php [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+            if (count($serializationGroups) > 0) {
+                $exclusion = new Exclusion($serializationGroups);
+            }
+
+            $list = new CollectionRepresentation($apiContacts, self::$entityKey, null, $exclusion, $exclusion);
+        }
+
+        $view = $this->view($list, 200);
+
+        // set serialization groups
```
Comment (turbo-ele) [inline Controller/ContactController.php]: format

File: Controller/ContactController.php [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+        return $this->container->getParameter('sulu.model.contact.class');
+    }
+
+    /**
+     * Applies the filter parameter and hasNoparent parameter for listbuilder.
+     *
+     * @param Request $request
+     * @param DoctrineListBuilder $listBuilder
+     */
+    protected function applyRequestParameters(Request $request, $listBuilder)
```
Comment (turbo-ele) [inline Controller/ContactController.php]: add types (for listbuilder)

File: Controller/ContactController.php (hunk 3 above)
Comment (stemod) [inline Controller/ContactController.php]: ???????

File: Controller/ContactController.php (hunk 5 above)
Comment (stemod) [inline Controller/ContactController.php]: ?????????

Comment (turbo-ele) [inline Controller/ContactController.php]:
```
 protected function applyRequestParameters(Request $request, DoctrineListBuilder $listBuilder)
```

File: Controller/ContactController.php (hunk 3 above)
Comment (turbo-ele) [inline Controller/ContactController.php]: i.m.o it should be 

```
// Convert to api-contacts.
```

At least a capitalized letter at the beginning would be nice ;)

File: Controller/ContactController.php [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    /**
+     * Takes an array of contacts and resets the avatar containing the media id with
+     * the actual urls to the avatars thumbnail.
+     *
+     * @param array $contacts
+     * @param string $locale
+     *
+     * @return array
+     */
+    private function addAvatars($contacts, $locale)
```
Comment (turbo-ele) [inline Controller/ContactController.php:270]:
```
private function addAvatars(array $contacts, $locale)
```

File: Resources/public/js/components/contacts/edit/main.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    Edit.prototype.constructor = Edit;
+    Edit.prototype.loadComponentData = function() {
+        var promise = $.Deferred();
+        baseEdit.loadComponentData.call(this).then(function(data) {
+            if (!!this.options.contactType) {
+                data.type = ContactHeader.getContactTypeIdByTypeName(this.options.contactType);
+            }
+            promise.resolve(data);
+        }.bind(this));
+        return promise;
```
Comment (turbo-ele) [inline Resources/public/js/components/contacts/edit/main.js:38]: add new line before return

File: Resources/public/js/components/contacts/list/main.js [hunk 8]
```diff
@@ -9,17 +9,101 @@
...
         baseList,
+        dataUrlAddition = '',
 
-    clickCallback = function(item) {
+        constants = {
+            datagridInstanceName: 'contacts'
+        },
+
+        /**
+         * Generates the configs for the tabs in the header
```
Comment (turbo-ele) [inline Resources/public/js/components/contacts/list/main.js]: .

File: Resources/public/js/components/contacts/list/main.js [hunk 9]
```diff
@@ -9,17 +9,101 @@
...
+         * @returns {object} tabs options
```
Comment (turbo-ele) [inline Resources/public/js/components/contacts/list/main.js:30]: add new line before @return

File: Resources/public/js/components/contacts/list/main.js [hunk 10]
```diff
@@ -9,17 +9,101 @@
...
+            return {
+                componentOptions: {
+                    callback: selectFilter.bind(this),
+                    preselector: 'position',
+                    preselect: preselect
+                },
+                data: items
+            };
+        },
+        addNewContact = function(typeName) {
```
Comment (turbo-ele) [inline Resources/public/js/components/contacts/list/main.js:92]: add new line before function

File: Resources/public/js/components/contacts/list/main.js [hunk 11]
```diff
@@ -9,17 +9,101 @@
...
+         * @returns {object} tabs options
+         */
+        getTabConfigs = function() {
+            var items, i, index, type,
+                contactTypes,
+                contactSection = AppConfig.getSection('sulu-contact-extension'),
+                contactType,
+                preselect;
+
+            // check if contactTypes exist
```
Comment (turbo-ele) [inline Resources/public/js/components/contacts/list/main.js]: reformat comments

File: Resources/public/js/components/contacts/list/main.js [hunk 12]
```diff
@@ -59,5 +143,79 @@ define([
...
+                data: tabConfigs.data,
+                componentOptions: tabConfigs.componentOptions
+            };
+        }
+        header.tabs = tabs;
+
+        var dropdownItems = [];
+        var contactTypes = AppConfig.getSection('sulu-contact-extension').contactTypes;
+        this.sandbox.util.each(contactTypes, function (index, value) {
+            console.log(this.sandbox.translate(value.addTranslation));
```
Comment (turbo-ele) [inline Resources/public/js/components/contacts/list/main.js]: remove

File: Resources/public/js/components/contacts/list/main.js [hunk 13]
```diff
@@ -59,5 +143,79 @@ define([
...
+            tabs = {
+                data: tabConfigs.data,
+                componentOptions: tabConfigs.componentOptions
+            };
+        }
+        header.tabs = tabs;
+
+        var dropdownItems = [];
+        var contactTypes = AppConfig.getSection('sulu-contact-extension').contactTypes;
+        this.sandbox.util.each(contactTypes, function (index, value) {
```
Comment (turbo-ele) [inline Resources/public/js/components/contacts/list/main.js:166]: i would add a new line before foreach

File: Resources/public/js/services/contact-header.js [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+define(['app-config'], function(AppConfig) {
+
+    'use strict';
+
+    // get contact type based on information given
```
Comment (turbo-ele) [inline Resources/public/js/services/contact-header.js]: guess what?

File: Resources/public/js/services/contact-header.js [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+        /**
+         * Returns items for header.
+         *
+         * @param {Object} contactTypes
+         * @param {String} key
+         *
+         * @returns {Object}
+         */
+        getHeaderItem = function(contactTypes, key) {
+
```
Comment (turbo-ele) [inline Resources/public/js/services/contact-header.js]: remove empty line

File: Resources/public/js/services/contact-header.js [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+         * Sets header toolbar with conversion options according to configuration.
+         */
+        setHeaderToolbar = function(contactType) {
+            if (!contactType.convertableTo || !Object.keys(contactType.convertableTo).length) {
+                this.sandbox.emit('sulu.header.toolbar.item.hide', 'settings');
+            }
+        };
+
+    return {
+
```
Comment (turbo-ele) [inline Resources/public/js/services/contact-header.js]: remove empty line

File: Resources/public/js/services/contact-header.js [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+        /**
+         * Sets header data: breadcrumb, headline and content tabs for contact.
+         *
+         * @param {Object} contact Backbone-Entity
+         * @param {String} [contactTypeName] Name of contact entity
```
Comment (turbo-ele) [inline Resources/public/js/services/contact-header.js]: remove

File: Resources/public/js/services/contact-router.js [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+            url += '/type:' + type;
+        }
+        Mediator.emit('sulu.router.navigate', url);
+    };
+
+    ContactRouter.getInstance = function() {
+        if (instance === null) {
+            instance = new ContactRouter();
+        }
+        return instance;
```
Comment (turbo-ele) [inline Resources/public/js/services/contact-router.js:44]: add new line before return

File: Resources/translations/sulu/backend.de.xlf [hunk 19]
```diff
@@ -18,6 +18,30 @@
...
+                <source>contact.contact.type.basic</source>
+                <target>Basis</target>
+            </trans-unit>
+            <trans-unit id="8" resname="contact.contact.add-basic">
+                <source>contact.contact.add-basic</source>
+                <target>Neuen Kontakt hinzufügen</target>
+            </trans-unit>
+            <trans-unit id="9" resname="contact.contact.add-customer">
+                <source>contact.contact.add-customer</source>
+                <target>Neuen Kunde-Kontakt hinzufügen</target>
```
Comment (turbo-ele) [inline Resources/translations/sulu/backend.de.xlf]: Neuen Kunden-Kontakt hinzufügen

**Comment by turbo-ele** [comment]
else ack

**Comment by turbo-ele** [comment]
else ack

**Pull request merged**